pkg/network/handlers/block_announcement.go

```diff
@@ -0,0 +1,506 @@
+			return
+		case msg := <-ba.sendCh:
+			if ba.stream == nil {
+				log.Println("(ba *BlockAnnouncer) sendLoop() - Stream is nil, cannot send message")
+				return
+			}
+
+			if err := WriteMessageWithContext(ba.ctx, ba.stream, msg); err != nil {
+				log.Println("(ba *BlockAnnouncer) sendLoop() - Failed to send message:", err)
+				return
+			}
+		}
+	}
+}
+
+// receiveLoop continuously reads messages from the stream and processes
+// them as block announcements. It terminates when an error occurs or
+// the context is canceled.
+func (ba *BlockAnnouncer) receiveLoop() {
+	for {
+		select {
+		case <-ba.ctx.Done():
+			return
+		default:
+			msg, err := ReadMessageWithContext(ba.ctx, ba.stream)
+			if err != nil {
+				log.Println("(ba *BlockAnnouncer) receiveLoop() - Failed to read message:", err)
+				return
+			}
+			if err := ba.processAnnouncement(msg.Content); err != nil {
+				log.Println("(ba *BlockAnnouncer) receiveLoop() - process announcement:", err)
+				return
+			}
+		}
+	}
+}
+
+// processAnnouncement handles a block announcement received from the peer.
+// It extracts the header and finalized block information, validates the announcement,
+// stores the new header, and requests the full block if needed.
+func (ba *BlockAnnouncer) processAnnouncement(content []byte) error {
+	if len(content) < 333 { // Minimum size: header(297) + finalized(36)
+		return fmt.Errorf("announcement message too short")
+	}
+
+	// Extract header
+	var header block.Header
+	peerFinalized := chain.LatestFinalized{}
+
+	err := jam.Unmarshal(content[:headerSize], &header)
+	if err != nil {
+		return fmt.Errorf("unmarshal header: %w", err)
+	}
+	copy(peerFinalized.Hash[:], content[headerSize:headerSize+32])
+
+	peerFinalized.TimeSlotIndex = jamtime.Timeslot(binary.LittleEndian.Uint32(content[headerSize+32:]))
+
+	// Validate the announcement: announced block must be after peer's finalized block
+	if header.TimeSlotIndex <= peerFinalized.TimeSlotIndex {
+		return fmt.Errorf("announced block slot (%d) not after peer's finalized slot (%d)", header.TimeSlotIndex, peerFinalized.TimeSlotIndex)
+	}
+
+	// Check if we know the peer's finalized block
+	_, err = ba.Store.GetHeader(peerFinalized.Hash)
+	if err != nil {
+		if errors.Is(err, store.ErrHeaderNotFound) {
+			// We don't know peer's finalized block - might need to sync
+			fmt.Printf("Warning: Peer's finalized block %x unknown", peerFinalized.Hash)
+		} else {
+			return fmt.Errorf("checking for peer's finalized block in our db: %w", err)
+		}
+	}
+
+	// Process the new header according to our chain rules
+	if err = ba.HandleNewHeader(&header); err != nil {
+		return fmt.Errorf("process new block: %w", err)
+	}
+
+	// Request the full block using the CE 128 protocol
+	h, err := header.Hash()
+	if err != nil {
+		return fmt.Errorf("hash header: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(ba.ctx, 6*time.Second) // Arbitrary timeout
```

Comment:
I suggest making the timeout a const to make it easier to configure in the future